plugins/pulp_rpm/plugins/distributors/yum/publish.py

```diff
@@ -633,57 +624,89 @@ def finalize(self):
+            # "long" name field is good to have.
+            'short': repo_id,
+
+            # Pulp injects this value to indicate the repository a given pkglist was imported
+            # *from*. Since the next step is to filter all packages available in this errata known
+            # to pulp by all packages known to pulp, this should be set to the repo_id being
+            # published in case it's expected to exist in later steps.
+            '_pulp_repo_id': repo_id,
+
+            # Related to the need for a unique collection name, errata should only contain one
+            # pkglist. While filtering packages to include in this errata, add them to this single
+            # list of packages. pkglists can contain multiple collections of packages, but limiting
+            # the published errata to one collection per pkglist and one pkglist per errata per
+            # repository makes the collection name uniqueness requirement easy to fulfill using
+            # the repo_id only.
+            'packages': [],
 
-        # If we can pull a repo_id off the conduit, use that to generate repo-specific nevra
-        if self.context.conduit and hasattr(self.context.conduit, 'repo_id'):
-            repo_unit_nevra = self.context.get_repo_unit_nevra(erratum_unit)
-        else:
-            repo_unit_nevra = None
+        }
 
-        seen_pkglists = set()
         for pkglist in erratum_unit.pkglist:
-            packages = tuple(sorted(p['filename'] for p in pkglist['packages']))
-            if packages in seen_pkglists:
-                continue
-            seen_pkglists.add(packages)
-
-            if repo_unit_nevra is None:
-                pkglists_to_publish.append(pkglist)
-                continue
-
-            packages_to_publish = []
-            for package in pkglist['packages']:
-                package_nevra = {'name': package['name'],
-                                 'version': package['version'],
-                                 'release': package['release'],
-                                 'epoch': package['epoch'] or '0',
-                                 'arch': package['arch']}
-
-                if package_nevra not in repo_unit_nevra:
-                    # current package not in the specified repo, don't add it to the pkglist
-                    continue
-
-                packages_to_publish.append(package)
-
-            if packages_to_publish:
-                new_pkglist = dict((key, val) for key, val in pkglist.items() if key != 'packages')
-                new_pkglist['packages'] = packages_to_publish
-                pkglists_to_publish.append(new_pkglist)
+            if repo_unit_nevra:
+                for package in pkglist.get('packages', []):
+                    package_nevra = models.NEVRA._fromdict(package)
+                    if package_nevra in repo_unit_nevra and package_nevra not in seen_packages:
```

Comment:
Checking `seen_packages` here is a good way to make packages unique in a `new_pkglist` though with such approach we no longer skip duplicate pkglists as a whole. And most likely there will be fully identical pkglists in DB rather than slightly different ones.
I am not proposing anything, just giving more context so you know why it was done that way.